### PR TITLE
Add Docker devcontainer configuration files

### PR DESCRIPTION
**Pull request by HManiac74**
Added configuration files for running the repo in a Docker dev Container

### CODE REVIEW

File: .devcontainer/devcontainer.json [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+				"python.linting.pylintEnabled": true,
+				"python.formatting.autopep8Path": "/usr/local/py-utils/bin/autopep8",
+				"python.formatting.blackPath": "/usr/local/py-utils/bin/black",
+				"python.formatting.yapfPath": "/usr/local/py-utils/bin/yapf",
+				"python.linting.banditPath": "/usr/local/py-utils/bin/bandit",
+				"python.linting.flake8Path": "/usr/local/py-utils/bin/flake8",
+				"python.linting.mypyPath": "/usr/local/py-utils/bin/mypy",
+				"python.linting.pycodestylePath": "/usr/local/py-utils/bin/pycodestyle",
+				"python.linting.pydocstylePath": "/usr/local/py-utils/bin/pydocstyle",
+				"python.linting.pylintPath": "/usr/local/py-utils/bin/pylint"
```
Comment (cclauss) [inline .devcontainer/devcontainer.json:33]: We use `ruff` to replace all of this.  Please ensure that `ruff` is working properly.
```suggestion
				"python.defaultInterpreterPath": "/usr/local/bin/python",
				"python.linting.enabled": true,
				"python.formatting.blackPath": "/usr/local/py-utils/bin/black",
				"python.linting.mypyPath": "/usr/local/py-utils/bin/mypy",
```

File: .devcontainer/devcontainer.json [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+			]
+		}
+	},
+
+	// Use 'forwardPorts' to make a list of ports inside the container available locally.
+	// "forwardPorts": [],
+
+	// Use 'postCreateCommand' to run commands after the container is created.
+	// "postCreateCommand": "pip3 install --user -r requirements.txt",
+
```
Comment (cclauss) [inline .devcontainer/devcontainer.json:49]: Add a \n to the end of both files to get rid of ⛔ as seen at https://github.com/TheAlgorithms/Python/pull/8887/files#

**Review by cclauss — CHANGES_REQUESTED** [review]
We already have GitPod...  Why is this any better than that?

We run on Python 3.11 so please upgrade from `bullseye` to `bookworm` and remove all comments about Python < 3.11.

Remove all the comments about Node.js because this repo has none of that.

File: requirements.txt [hunk 3]
```diff
@@ -11,6 +11,7 @@ projectq
 qiskit
 requests
 rich
+ruff
```
Comment (cclauss) [inline requirements.txt:14]: Not needed.  Please remove.  There is a difference between runtime dependencies and testing dependencies.

File: .devcontainer/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+ARG VARIANT=3.11-bookworm
+FROM mcr.microsoft.com/vscode/devcontainers/python:${VARIANT}
+
+# [Choice] Node.js version: none, lts/*, 16, 14, 12, 10
+# ARG NODE_VERSION="none"
+# RUN if [ "${NODE_VERSION}" != "none" ]; then su vscode -c "umask 0002 && . /usr/local/share/nvm/nvm.sh && nvm install ${NODE_VERSION} 2>&1"; fi
+
+# [Optional] If your pip requirements rarely change, uncomment this section to add them to the image.
+COPY requirements.txt /tmp/pip-tmp/
+RUN pip3 --disable-pip-version-check --no-cache-dir install -r /tmp/pip-tmp/requirements.txt
```
Comment (cclauss) [inline .devcontainer/Dockerfile:11]:
```suggestion
RUN pip3 --disable-pip-version-check --no-cache-dir install ruff -r /tmp/pip-tmp/requirements.txt
```

File: .devcontainer/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+# [Optional] Uncomment this section to install additional OS packages.
+# RUN apt-get update && export DEBIAN_FRONTEND=noninteractive \
+#     && apt-get -y install --no-install-recommends <your-package-list-here>
+
+# [Optional] Uncomment this line to install global node packages.
+# RUN su vscode -c "source /usr/local/share/nvm/nvm.sh && npm install -g <your-package-here>" 2>&1
```
Comment (cclauss) [inline .devcontainer/Dockerfile]:
```suggestion
# https://github.com/microsoft/vscode-dev-containers/blob/main/containers/python-3/README.md
ARG VARIANT=3.11-bookworm
FROM mcr.microsoft.com/vscode/devcontainers/python:${VARIANT}
COPY requirements.txt /tmp/pip-tmp/
RUN python3 -m pip install --upgrade pip \
  && python3 -m pip install --no-cache-dir install ruff -r /tmp/pip-tmp/requirements.txt
```

File: .devcontainer/devcontainer.json [hunk 6]
```diff
@@ -0,0 +1,44 @@
+{
+	"name": "Python 3",
+	"build": {
+		"dockerfile": "Dockerfile",
+		"context": "..",
+		"args": {
+			// Update 'VARIANT' to pick a Python version: 3, 3.10, 3.9, 3.8, 3.7, 3.6
+			// Append -bullseye or -buster to pin to an OS version.
+			// Use -bullseye variants on local on arm64/Apple Silicon.
```
Comment (cclauss) [inline .devcontainer/devcontainer.json:9]:
```suggestion
```

File: .devcontainer/devcontainer.json [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+			"VARIANT": "3.11-bookworm",
+			// Options
+			"NODE_VERSION": "lts/*"
```
Comment (cclauss) [inline .devcontainer/devcontainer.json:12]:
```suggestion
```

File: .devcontainer/devcontainer.json [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+				"python.linting.enabled": true,
+				"python.formatting.blackPath": "/usr/local/py-utils/bin/black",
+				"python.linting.mypyPath": "/usr/local/py-utils/bin/mypy"
+			},
+
+			// Add the IDs of extensions you want installed when the container is created.
+			"extensions": [
+				"ms-python.python",
+				"ms-python.vscode-pylance"
+			]
```
Comment (cclauss) [inline .devcontainer/devcontainer.json:32]: In a **future** PR, it would be cool to add... https://github.com/astral-sh/ruff-vscode

**Review by cclauss — CHANGES_REQUESTED** [review]

File: .devcontainer/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,6 @@
+# https://github.com/microsoft/vscode-dev-containers/blob/main/containers/python-3/README.md
+ARG VARIANT=3.11-bookworm
+FROM mcr.microsoft.com/vscode/devcontainers/python:${VARIANT}
+COPY requirements.txt /tmp/pip-tmp/
+RUN python3 -m pip install --upgrade pip \
```
Comment (cclauss) [inline .devcontainer/Dockerfile:5]: 🚫

**Review by CaedenPH — CHANGES_REQUESTED** [review]
As cclauss mentioned, we have a gitpod config file, and I don't see much value in adding devcontainer config files, primarily due to the fact that this is unlike other repositories, and that it is a group of standalone files that don't require much, if any config
This also only customises for a singular tool, vscode, excluding other text editors/IDE's
Even still, if this is decided to be merged, perhaps it would be ideal to add information on how to set up this container in the `CONTRIBUTING.md`

**Comment by HManiac74** [comment]
Some people might prefer a local development container, and VSCode is a very good IDE, and platform independent, so I thought this might be a good idea

**Review by cclauss — APPROVED** [review]
Let's see if this AND GitPod can cohabitate without nullifying each other.  Contributors can choose their favorite.

Let's reserve the right to revert this contribution if users protest or if it does not bring sufficient value.

@CaedenPH is correct that we need docs in CONTRIBUTING.md so @HManiac74 please create a follow-on PR that helps folks to get started.

**Pull request merged**